exercise/cache/content.py

```diff
@@ -183,6 +185,44 @@ def add_to(target: Dict[str, Any], exercise: BaseExercise) -> None:
         if total['min_group_size'] > total['max_group_size']:
             total['min_group_size'] = 1
 
+        exercises = []
+        for module in modules:
+            #add the module fields to cached exercise endpoint
+            api_url = f"{settings.BASE_URL}/api/v2/courses/{instance.id}/exercises/{module['id']}/"
+            module_html_url = f"{settings.BASE_URL}{module['link']}"
+            m = {
+                'id': module['id'],
+                'url': api_url,
+                'html_url': module_html_url,
+                'display_name': module['name'],
+                'is_open': module['is_open'],
+                'exercises': []
+            }
+            #add this module to cached exercises
+            exercises.append(m)
+            for submodule in module['children']:
+                children = []
+                #loop over each exercise in the module
+                for exercise in filter(lambda ex: ex['type'] == 'exercise', submodule['children']):
+                    #extract all the relevant fields of the exercise
+                    api_url = f"{settings.BASE_URL}/api/v2/exercises/{exercise['id']}/"
+                    html_url = f"{module_html_url}{exercise['get_path']}/"
+                    ed = {
+                            'id': exercise['id'],
+                            'url': api_url,
+                            'html_url': html_url,
+                            'display_name': exercise['name'],
+                            'max_points': exercise['max_points'],
+                            'max_submissions': exercise['max_submissions'],
+                            'hierarchical_name': exercise['name'],
+                            'difficulty': exercise['difficulty']
+                    }
+                    #add this exercise to the parent module
+                    exercises[-1]['exercises'].append(ed)
+
+        #remove the temporary is_open field from modules used in exercises
+        for module in modules:
+            module.pop('is_open', None)
```

Comment:
A similar for-loop already exists above, so this line should be moved there. That way all modules don't have need to be looped over again.

Reply:
So the reason I decided to iterate over the modules list rather than this loop (which I'm assuming that you're referring to)
https://github.com/EerikSaksi/a-plus/blob/cd6deb3b9d7347efc7793837780a2fc9da9ba0ac/exercise/cache/content.py#L104

is because the module in the scope of this loop does not have access to the exercises that belong to this module. This loop edits the module list and adds the corresponding exercises to each module. I could try to build the data in this loop a well but looping through the structured data felt far easier than building it through the recursion in the loop.